File: cs/Minsk/CodeAnalysis/Binding/BoundUnaryOperator.cs
using Minsk.CodeAnalysis.Syntax;

namespace Minsk.CodeAnalysis.Binding;

internal sealed class BoundUnaryOperator
{
    private static readonly BoundUnaryOperator[] KnownOperators =
    {
        new(SyntaxKind.PlusToken, BoundUnaryOperatorKind.Identity, typeof(int), typeof(int)),
        new(SyntaxKind.MinusToken, BoundUnaryOperatorKind.Negation, typeof(int), typeof(int)),
        new(SyntaxKind.BangToken, BoundUnaryOperatorKind.LogicalNegation, typeof(bool), typeof(bool)),
        new(SyntaxKind.TildeToken, BoundUnaryOperatorKind.BitwiseNegation, typeof(int), typeof(int))
    };

    private BoundUnaryOperator(
        SyntaxKind syntaxKind,
        BoundUnaryOperatorKind operatorKind,
        Type operandType,
        Type resultType
    )
    {
        SyntaxKind = syntaxKind;
        OperatorKind = operatorKind;
        OperandType = operandType;
        ResultType = resultType;
    }

    public BoundUnaryOperatorKind OperatorKind { get; }
    private SyntaxKind SyntaxKind { get; }
    private Type OperandType { get; }
    public Type ResultType { get; }

    public static BoundUnaryOperator? Bind(SyntaxKind operatorKind, BoundExpression operand)
    {
        return KnownOperators.FirstOrDefault(op => op.SyntaxKind == operatorKind && op.OperandType == operand.Type);
    }
}

File: kt/minsk/src/main/kotlin/dev/phytolizer/minsk/analysis/binding/BoundUnaryOperator.kt
package dev.phytolizer.minsk.analysis.binding

import dev.phytolizer.minsk.analysis.syntax.SyntaxKind
import kotlin.reflect.KClass

internal class BoundUnaryOperator private constructor(
    private val syntaxKind: SyntaxKind,
    val kind: BoundUnaryOperatorKind,
    private val operandType: KClass<out Any>,
    val resultType: KClass<out Any>,
) {
    companion object {
        private val operators = listOf(
            BoundUnaryOperator(SyntaxKind.PlusToken, BoundUnaryOperatorKind.Identity, Int::class, Int::class),
            BoundUnaryOperator(SyntaxKind.MinusToken, BoundUnaryOperatorKind.Negation, Int::class, Int::class),
            BoundUnaryOperator(
                SyntaxKind.BangToken,
                BoundUnaryOperatorKind.LogicalNegation,
                Boolean::class,
                Boolean::class,
            ),
        )

        fun bind(syntaxKind: SyntaxKind, operandType: KClass<out Any>): BoundUnaryOperator? {
            return operators.firstOrNull { it.syntaxKind == syntaxKind && it.operandType == operandType }
        }
    }
}
